### PR TITLE
GH-36994: [Java] Use JDK 21 in CI

### LINKED ISSUE #36994: [Java][CI] Enable JDK 21
### Describe the enhancement requested

Java 21 is the next long-term support (LTS) version after Java 17 and is scheduled to be released in September 2023. 

### Component(s)

Java

### PR DESCRIPTION
**Pull request by danepitkin**
### Rationale for this change

Verify JDK 21 in CI in time for the Arrow v14 release.

### What changes are included in this PR?

* Bump latest Java version from 20 -> 21 in CI

### Are these changes tested?

Yes, via CI.

### Are there any user-facing changes?

No.
* Closes: #36994

### CODE REVIEW

**Comment by danepitkin** [comment]
Eclipse temurin released JDK 21 docker images, but the maven integration is not available yet. https://hub.docker.com/_/maven

**Comment by danepitkin** [comment]
@github-actions crossbow submit *java*

File: java/flight/pom.xml [hunk 1]
```diff
@@ -52,4 +52,17 @@
             </plugins>
         </pluginManagement>
     </build>
+
+    <profiles>
```
Comment (raulcd) [inline java/flight/pom.xml:56]: does this change mean we can potentially remove this pinned version for release verification?
https://github.com/apache/arrow/blob/main/dev/release/verify-release-candidate.sh#L569

**Review by raulcd — APPROVED** [review]
Only a minor comment but I am happy to merge as is and (possibly) update the release verification script, if necessary, on a separate PR

File: java/flight/pom.xml (hunk 1 above)
Comment (danepitkin) [reply java/flight/pom.xml:56]: I think so! I'll try it out

**Comment by danepitkin** [comment]
@github-actions crossbow submit *java*

**Comment by danepitkin** [comment]
@github-actions crossbow submit verify-rc-source-java-macos-amd64

**Review by raulcd — APPROVED** [review]
:+1:

**Pull request merged**